mesonbuild/mintro.py

```diff
@@ -125,7 +131,160 @@ def list_target_files(target_name, coredata, builddata):
+
+    def add_languages(self, args):
+        need_cross_compiler = self.environment.is_cross_build() and self.environment.cross_info.need_cross_compiler()
+        for lang in sorted(args, key=compilers.sort_clink):
+            lang = lang.lower()
+            if lang in self.coredata.compilers:
+                comp = self.coredata.compilers[lang]
+            else:
+                (comp, _) = self.detect_compilers(lang, need_cross_compiler)
+            if comp is None:
+                return False
+            self.add_base_options(comp)
+        return True
+
+    def func_project(self, node, args, kwargs):
+        if len(args) < 1:
+            raise InvalidArguments('Not enough arguments to project(). Needs at least the project name.')
+
+        proj_langs = self.flatten_args(args[1:])
+
+        if os.path.exists(self.option_file):
+            oi = optinterpreter.OptionInterpreter(self.subproject)
+            oi.process(self.option_file)
+            self.coredata.merge_user_options(oi.options)
+
+        def_opts = kwargs.get('default_options', [])
+        if isinstance(def_opts, mparser.ArrayNode):
+            def_opts = [x.value for x in def_opts.args.arguments]
+
+        self.project_default_options = mesonlib.stringlistify(def_opts)
+        self.project_default_options = cdata.create_options_dict(self.project_default_options)
+        self.default_options.update(self.project_default_options)
+
+        if not self.is_subproject() and 'subproject_dir' in kwargs:
+            spdirname = kwargs['subproject_dir']
+            if isinstance(spdirname, str):
+                self.subproject_dir = spdirname
+        if not self.is_subproject():
+            subprojects_dir = os.path.join(self.source_root, self.subproject_dir)
+            if os.path.isdir(subprojects_dir):
+                for i in os.listdir(subprojects_dir):
+                    if os.path.isdir(os.path.join(subprojects_dir, i)):
+                        self.do_subproject(i)
+
+        self.coredata.init_backend_options(self.backend.name)
+        options = {k: v for k, v in self.environment.cmd_line_options.items() if k.startswith('backend_')}
+
+        self.coredata.set_options(options)
+        self.add_languages(proj_langs)
+
+    def do_subproject(self, dirname):
+        subproject_dir_abs = os.path.join(self.environment.get_source_dir(), self.subproject_dir)
+        subpr = os.path.join(subproject_dir_abs, dirname)
+        try:
+            subi = BuildoptionsInterperter(subpr, '', self.backend.name, cross_file=self.cross_file, subproject=dirname, subproject_dir=self.subproject_dir)
+            subi.subproject_stack = self.subproject_stack + [dirname]
+            subi.analyze()
+        except:
+            # Meh, can not log this because of JSON output
+            return
+
+        self.coredata.user_options.update(subi.coredata.user_options)
+        self.coredata.compiler_options.update(subi.coredata.compiler_options)
```

Comment:
These two lines are needed because each interpreter has it's own environment and thus coredata, right?

Reply:
Yes. And having a shared `coredata` and / or `environment` may break stuff when the subproject modifies entries from the root project.

Reply:
How could that be a problem when the real interpreter does share its environment and coredate as well. (via `self.environment = build.environment` and `self.coredata = self.environment.get_coredata()` in Interpreter.__init__

Reply:
It doesn't: `Coredata` is stored in `Environment` which is stored in `Build`:
https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreter.py#L1948-L1950

And for each subproject a _copy_ of `Build` is created, which is later merged:
https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreter.py#L2381

I don't exactly know what happens during the merge process and how that affects the build options.

Reply:
Yes, but that copy does not actually deep copy environment. It only copies OrderedDict·s. All other attributes of Build are just reference copied.

Reply:
Oh yes, I didn't look at `copy()` itself. But isn't this a bug? I mean, what happens when the subproject fails and is not required?

It should be fin for the introspection, however (unless it breaks my test case).